tempesta_fw/http_msg.c

```diff
@@ -475,21 +475,65 @@ __hdr_sub(TfwHttpMsg *hm, char *name, size_t n_len, char *val, size_t v_len,
 			{ .ptr = name,	.len = n_len },
 			{ .ptr = ": ",	.len = 2 },
 			{ .ptr = val,	.len = v_len },
-			{ .ptr = "\r\n", .len = 2 }
+			{ .ptr = "\r\n",.len = 2 } /* may be adjusted */
 		},
 		.len = n_len + v_len + 4,
 		.flags = 4 << TFW_STR_CN_SHIFT
 	};
+	TfwStr *hlc = TFW_STR_LAST(&hdr);
 
-	if (!TFW_STR_DUP(orig_hdr) && hdr.len <= orig_hdr->len) {
-		/* Rewrite the header in-place. */
-		if (ss_skb_cutoff_data(orig_hdr, hdr.len, 2))
-			return TFW_BLOCK;
-		if (tfw_strcpy(orig_hdr, &hdr))
-			return TFW_BLOCK;
-		return 0;
+	if (TFW_STR_DUP(orig_hdr))
+		goto slow_path;
+
+	BUG_ON(!tfw_str_eolen(orig_hdr));
+
+	/*
+	 * We are trying to keep line endings as they were received, so
+	 * adjustment of @hdr's last chunk is needed in case of single LF.
+	 */
+	if (tfw_str_eolen(orig_hdr) == 1)
+		++hlc->ptr, --hlc->len, --hdr.len;
+
+	/*
+	 * In general, there are 5 possible layouts for @hdr and @orig_hdr. And
+	 * in case of single LF used as EOL the 3-rd case is degenerated.
+	 *
+	 *   orig_hdr     | Header: Value | EOL    =  < LF | CRLF >
+	 *   ------------- --------------- ---------------------------------
+	 *
+	 *        hdr     | Foo: BarEOL.. |        1) cutoff, fast_path
+	 *        hdr     | Foo: ..BarEOL |        2) cutoff, fast_path
+	 *
+	 *        hdr     | Foo: ...BarEO | L      3) slow_path (CRLF only)
+	 *        hdr     | Foo: .....Bar | EOL    4) reuse EOL, fast_path
+	 *
+	 *        hdr     | Foo: ......Ba | rEOL   5) slow_path
+	 */
+	if (hdr.len > orig_hdr->len) {
```

Comment:
There's a lots of case listed here. In my opinion this only clouds the issue.

The pre-condition here is that EOL bytes for both the original string and the substitute string are normalized and made the same. So the real condition for the fast path is that the length of the substitute string without the EOL bytes must be no larger than the length of the original string which is already counted without the EOL bytes. That is the only case when we can use the fast path. Moreover, if the length of the substitute string without the EOL bytes is exactly the same as the length of the original string (already without the EOL bytes), then we don't need to cut off any bytes. We can just copy the substitute string over the original string leaving the EOL bytes intact.

In all other cases we must go for the slow path.

Here is a sample code that expesses this logic in a clear and obvious way. It also has less branches (goto). What do you guys think?


```C
/* The same for both strings. */
eolen = tfw_str_eolen(orig_hdr);

/*
 * EOL bytes are not a part of a header field string in Tempesta.
 * Therefore only @orig_hdr->len bytes at most can be copied over.
 * The EOL has been normalized to be the same for both strings.
 * If the substitute string without the EOL fits into that space,
 * then the fast path can be used. Otherwise, go by the slow path.
 */
hdrlen_wo_eol = hdr.len - eolen;
if (hdrlen_wo_eol <= orig_hdr->len) {
        if (hdrlen_wo_eol == orig_hdr->len) {
                hdr.len -= eolen;
                TFW_STR_CHUNKN_SUB(&hdr, 1);
                goto copy_over;
        }
        if (ss_skb_cutoff_data(orig_hdr, hdr.len, eolen))
                return TFW_BLOCK;
copy_over:
        return tfw_strcpy(orig_hdr, &hdr) ? TFW_BLOCK : 0;
}
/* Generic and slower path. */
```